### PR TITLE
[CI] Actions - Windows mswin - use pre-built vcpkg, update to OpenSSL 3

### PR DESCRIPTION
**Pull request by MSP-Greg**
Adds using a vcpkg 7z archive which contains all the packages needed for Ruby (libffi, libyaml, openssl, readline, & zlib).

See https://bugs.ruby-lang.org/issues/18689

~~Note that all unchanged workflows were disabled...~~

See the first two commits (https://github.com/ruby/ruby/pull/5792/commits/839626a8218a and https://github.com/ruby/ruby/pull/5792/commits/a80d908eb6a8)~~, the last commit is just for the draft PR~~.

### CODE REVIEW

File: ext/openssl/extconf.rb [hunk 1]
```diff
@@ -184,7 +184,7 @@ def find_openssl_library
 
 Logging::message "=== Checking done. ===\n"
 
-if openssl_3
+if openssl_3 && !$mswin
```
Comment (nobu) [inline ext/openssl/extconf.rb:187]: Is this needed?
`$warnflags` should not have this flag on mswin builds, I think.

Comment (MSP-Greg) [reply ext/openssl/extconf.rb:187]: I can check again, but as I recall `$warnflags` was nil, so on the next line `$warnflags.sub!` throws?

File: .github/workflows/windows.yml [hunk 2]
```diff
@@ -44,23 +44,16 @@ jobs:
...
-            ${{ runner.os }}-vcpkg-installed-${{ matrix.os }}-
-            ${{ runner.os }}-vcpkg-installed-
-      - name: Install libraries with vcpkg
+      - name: Install pre-built vcpkg libraries
         run: |
-          vcpkg --triplet x64-windows install libffi libyaml openssl readline zlib
+          cd $env:VCPKG_INSTALLATION_ROOT
+          $uri = "https://github.com/MSP-Greg/setup-msys2-gcc/releases/download/msys2-gcc-pkgs/mswin.7z"
+          curl -L $uri -o $env:RUNNER_TEMP/mswin_temp.7z
+          &"C:\Program Files\7-Zip\7z" x $env:RUNNER_TEMP/mswin_temp.7z -aoa
```
Comment (nobu) [inline .github/workflows/windows.yml:52]: I didn’t know this usage of `&`.

File: ext/openssl/extconf.rb (hunk 1 above)
Comment (nobu) [reply ext/openssl/extconf.rb:187]: Indeed, it’s forgotten in win32/makefile.sub.
Then isn’t it better to check that variable itself?

Comment (nobu) [reply ext/openssl/extconf.rb:187]:
```suggestion
if openssl_3 && $warnflags
```

Comment (MSP-Greg) [reply ext/openssl/extconf.rb:187]: I'll defer to your judgement.  Wasn't sure if it should be changed in mkmf.rb or elsewhere...

Comment (MSP-Greg) [reply ext/openssl/extconf.rb:187]: Good suggestion.  I'll update.

**Comment by MSP-Greg** [comment]
The change to [ext/openssl/extconf.rb](https://github.com/ruby/ruby/pull/5792/files#diff-bdb47184d401be1228fb9d4487514a1bdc08f13ca434b4218c8cb3d84f5624e1) may be best done in another commit or in ruby/openssl.  Will change here as needed.

A benefit of using vcpkg or mingw/ucrt packages is that they are often using newer versions of various dependencies than standard Ubuntu or macOS.  For instance, vcpkgs libffi is 3.4.2, Ubuntu 20.04 lists 3.3, which may be patched...

**Comment by MSP-Greg** [comment]
Added a commit, [MSP-Greg/setup-msys2-gcc -> ruby/setup-msys2-gcc](https://github.com/ruby/ruby/pull/5792/commits/eedac3e88995755d2b87b23d9dea4d1ac7d71d11)

**Pull request closed without merging**